### ISSUE TITLE
Where to define borgbackup_repository_storage?

### ISSUE THREAD

**Issue by indibart** [apollo13/ansible-role-borgbackup#1]
Hi there! Just tried your ansible role. Unfortunally it's missing the variable borgbackup_repository_storage. Where do I have to define it? I can't find the variable.
Cheers!

**Comment by apollo13** [comment]
Hi, borgbackup_repository_storage should be only defined on your remote server (assuming you are also managing it with ansible): https://github.com/apollo13/ansible-role-borgbackup/blob/master/README.md#server-variables

You'd just put it in host_vars/your_server.yml -- if that doesn't work for you please show me your inventory, playbook and the error you get.

**Comment by indibart** [comment]
Hi apollo, thanks that helped and I just learned a bit about Ansible. Still learning.
And I'm kind of stuck now, because the client side wont properly config. Can you help?

This is the error I get (laptop-001 ist the client-machine):

```
TASK [apollo13.borgbackup : Create borg config & temporary data directories] *************************************************
failed: [laptop-001] (item=config) => {"ansible_loop_var": "item", "changed": false, "gid": 0, "group": "root", "item": "config", "mode": "0755", "msg": "chown failed: failed to look up user borg", "owner": "root", "path": "/home/borg/config", "size": 4096, "state": "directory", "uid": 0}
failed: [laptop-001] (item=data) => {"ansible_loop_var": "item", "changed": false, "gid": 0, "group": "root", "item": "data", "mode": "0755", "msg": "chown failed: failed to look up user borg", "owner": "root", "path": "/home/borg/data", "size": 4096, "state": "directory", "uid": 0}
```

My group_var/Computer.yml (laptop-001 ist part of the group Computer):
```
---
# Minimal-Konfiguration der Borg-Clients
borgbackup_repository_server: backupi-001

# Zusätzliche Konfigurationsmöglichkeiten der Borg-Clients
#borgbackup_passphrase: XXX_SECRET_XXX
borgbackup_calendar_spec: "*-*-* 11:16:00" # default, pattern is in systemd-timer format
#borgbackup_directories: ["{{ borgbackup_home }}/data"] # default
#borgbackup_exclude_patterns: [] # default, see borg help patterns, uses fnmatch-style format
#borgbackup_append_only: yes # default
```

**Comment by apollo13** [comment]
Your error message tells you:
```
chown failed: failed to look up user borg
```
As written down in the README.md of this repository, this role does not create the users it uses (I am not yet 100% sure if that is a good idea or not, but that is how it is for now).

So before calling my role you need something along the lines of:
```
- name: Create user for borgbackup
  user:
    name: "{{ borgbackup_user }}"
    state: present
```
I'll make sure to highlight this also in the readme, because that is easy to overlook.

**Comment by indibart** [comment]
I actually did set up the user borg but only for the server. I didn't realize that the client must also have the user.

New error though... And again I can't make sense of it. Can you help me again? Sorry!

```
TASK [apollo13.borgbackup : Init backup repository] **********************************************************************************************************************************************************************************************************
fatal: [pirislaptop001]: FAILED! => {"changed": false, "cmd": ["/home/borg/config/borg.sh", "init", "--encryption", "repokey-blake2"], "delta": "0:00:00.405864", "end": "2019-12-02 18:29:05.421818", "failed_when_result": true, "msg": "non-zero return code", "rc": 2, "start": "2019-12-02 18:29:05.015954", "stderr": "Remote: ssh: Could not resolve hostname backupi001: Name or service not known\r\nConnection closed by remote host. Is borg working on the server?", "stderr_lines": ["Remote: ssh: Could not resolve hostname backupi001: Name or service not known", "Connection closed by remote host. Is borg working on the server?"], "stdout": "", "stdout_lines": []}
```

**Comment by apollo13** [comment]
The role tries to use `backupi001` as a server. From the looks of it ansible reported it as fully qualified name for your system. As such it literally tries to connect to `backupi001`. What happens if you execute `ping backupi001` on your client? If you also get errors their your network is not configured properly and hosts are not reachable via their names.

The only option you have then is to set `borgbackup_repository/borgbackup_known_hosts` manually and to the IP of the host I guess. (Or you could set the hostname of your server to match your DNS)

**Comment by apollo13** [comment]
Hi @indibart, did you manage to fix your issues?

**Comment by indibart** [comment]
Hey @apollo13, sorry I didn't have time to look into it. Busy week.
Yes I fixed it, it had something to do with the network configuration of the VMs. Thanks for the hint!
But of course, two step forward, one step back. Another error:

`fatal: [pirislaptop001]: FAILED! => {"changed": false, "cmd": ["/home/borg/config/borg.sh", "init", "--encryption", "repokey-blake2"], "delta": "0:00:00.762230", "end": "2019-12-07 09:24:37.463687", "failed_when_result": true, "msg": "non-zero return code", "rc": 2, "start": "2019-12-07 09:24:36.701457", "stderr": "Remote: usage: borg [-h] [-V] <command> ...\nRemote: borg: error: unrecognized arguments: --restrict-to-repository /home/borg/pirislaptop001\nConnection closed by remote host. Is borg working on the server?", "stderr_lines": ["Remote: usage: borg [-h] [-V] <command> ...", "Remote: borg: error: unrecognized arguments: --restrict-to-repository /home/borg/pirislaptop001", "Connection closed by remote host. Is borg working on the server?"], "stdout": "", "stdout_lines": []}
`

I had a hunch that it could be because the client and the server didn't have the same version of borg installed. The client is a Linux Mint 19.2 machine while the server is a Debian 9 machine. I figured that the version of borg on Debian 9 is to old for --restrict-to-repository.

So I tested using the upstream installation method. That did solve the problem but gave an other error where I'm now stuck. Again.

`fatal: [pirislaptop001]: FAILED! => {"changed": false, "cmd": ["/home/borg/config/borg.sh", "init", "--encryption", "repokey-blake2"], "delta": "0:00:09.019260", "end": "2019-12-07 18:35:17.784959", "failed_when_result": true, "msg": "non-zero return code", "rc": 2, "start": "2019-12-07 18:35:08.765699", "stderr": "Remote: sh: 1: /usr/bin/borg: not found\nConnection closed by remote host. Is borg working on the server?", "stderr_lines": ["Remote: sh: 1: /usr/bin/borg: not found", "Connection closed by remote host. Is borg working on the server?"], "stdout": "", "stdout_lines": []}
`

Could it be that the upstream version of borg puts the borg app on a different place on the filesystem?

**Comment by indibart** [comment]
Yes! That was it!
On the server:
`$ whereis borg`
`$ borg: /usr/local/bin/borg`

So I created a link: `ln -s /usr/local/bin/borg /usr/bin/borg`
And it worked! 

I'm just not sure if that solution is a good idea or if it will break something in the future. What do you think?

**Comment by apollo13** [comment]
I have to look through the role and think about why this could happen. I thought I had that covered. I'll take a look tomorrow.

**Comment by indibart** [comment]
Sure! Thanks!

**Comment by apollo13** [comment]
hi @indibart, I tried to reproduce your issue but couldn't. Do you have any more information for me? Preferably a reproducible setup? The thing is that my current tests are covering your usecase, so I am somewhat out of ideas.

**Comment by apollo13** [comment]
The only thing I could think of is that you manually configured `borgbackup_executable` during your tests which now causes the issues on the server.

**Comment by indibart** [comment]
Sorry, didn't have time to look into it. Decembre is always a really busy time...

I don't think I did configure `backbackup_executable` in any way. But as I was playing with your role a bit it could be that I changed something without remembering it anymore.

Client and Server are VMs in Virtualbox and apart of changing the repos to local - but still official - ones, upgrading all the packages and exchanging ssh-keys of each other I didn't change anything... They are for testing purpose only.


Well, I will try a fresh start when I have time. Maybe on Sunday.
